nemo_text_processing/g2p/data/heteronym_classification_data.py

```diff
@@ -65,98 +67,136 @@ def __init__(
+        for homograph_, start_end_ in zip(homographs, start_end):
+            if homograph_.lower() != sentence[start_end_[0] : start_end_[1]].lower():
+                return None
+
+        # add bos token
+        input_ids = [self.tokenizer.bos_id]
+        subtokens_mask = [self.PAD_TOKEN]  # the first tokens of heteronym spans are 1s, the rest of the tokens are 0s
 
         if self.with_labels:
-            if word_id is None:
-                raise ValueError(f"word_id must be provided when self.with_labels==True, i.e., training mode")
-
-            target_word_id = self.wordid_to_idx[word_id]
-            # add extra -100 tokens at the begging and end for [CLS] and [SEP] bert tokens
-            target_word_id = (
-                [-100] * (l_context_len + 1)
-                + [target_word_id]
-                + [-100] * (homograph_len - 1)
-                + [-100] * (r_context_len + 1)
-            )
-            assert len(target_word_id) == len(self.tokenizer.tokenizer([sentence])["input_ids"][0])
-            output.append(target_word_id)
-
-        return output  # [target_and_negatives, subword_mask] and (optional) target
+            target_word_ids = [self.LOSS_PAD_TOKEN]  # -100 to pad plain tokens
+        else:
+            target_word_ids = None  # for inference when labels are not available
+
+        heteronym_span_idx = 0
+        # split sentence by space and keep track of word boundaries
+        # we assume heteronym is a standalone word
+        matches = [(m.group(0), (m.start(), m.end() - 1)) for m in re.finditer(r'\S+', sentence)]
+        for match in matches:
+            word, word_start_end = match
+            # check if the start of the next heteronym span is within the word indices
+            if (
+                heteronym_span_idx < len(start_end)
+                and word_start_end[0] <= start_end[heteronym_span_idx][0] < word_start_end[1]
+            ):
+                heteronym_start_end = start_end[heteronym_span_idx]
+                prefix = ""
+                prefix_ids = []
+                # for cases when word also includes punctuation marks at the beginning or a prefix,
+                # e.g. "diffuse" vs. diffuse vs. pre-diffuse for heteronym {diffuse}
+                if word_start_end[0] < heteronym_start_end[0]:
+                    prefix = sentence[word_start_end[0] : heteronym_start_end[0]]
+                    prefix_ids = self.tokenizer.text_to_ids(prefix)
+                    subtokens_mask.extend([self.PAD_TOKEN] * len(prefix_ids))
+
+                word = word[word.index(prefix) + len(prefix) :]
+                word_input_ids = self.tokenizer.text_to_ids(word)
+                input_ids.extend(prefix_ids + word_input_ids)
+
+                subtokens_mask.extend([1] + [self.PAD_TOKEN] * (len(word_input_ids) - 1))
+
+                if self.with_labels:
+                    cur_target_word_id = self.wordid_to_idx[word_ids[heteronym_span_idx]]
```

Comment:
i am not convinced that `[wordid_to_idx: Dict[str, int],]`(https://github.com/NVIDIA/NeMo/blob/be596f44240be7d7e3faa63e357189846e1e1c27/nemo_text_processing/g2p/data/heteronym_classification_data.py#L37) as defined in class `HeteronymClassificationDataset` But `[word_ids: Optional[List[int]] = None,]`(https://github.com/NVIDIA/NeMo/blob/be596f44240be7d7e3faa63e357189846e1e1c27/nemo_text_processing/g2p/data/heteronym_classification_data.py#L109). 

So the key for `self.wordid_to_idx` should be a `str`, but got `int`?

Reply:
right, line 109 should be List[str]. fixed